feat: Turn `RenderingContext` into a trait

dklassic opened the pull request:
<!-- Please describe your changes on the following line: -->

Turn `RenderingContext` into a trait and rename previous struct implementation as `SurfmanRenderingContext`.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #34888

<!-- Either: -->
- [x] These changes do not require tests because there's no behavior change

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

dklassic commented:
Would be nice if we can use NativeContext/NativeDevice but for now their implementation still differs between platforms. So we'll probably have to settle with this approach for now.

jdm commented:
My [embedding example](https://github.com/jdm/servo-embedding-example/) relies on the with_front_buffer API.

wusyong commented:
Hmm, I was thinking external GL clients will handle those buffers themselves. Let's just use surfman device and context for now then. I'll see if @jdm's approach can apply to WebGL's image handler later.

mrobinson commented on ports/servoshell/rendering_context.rs:
```diff
@@ -0,0 +1,297 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Let's keep the implementation of this `RenderingContext` the `webrender` crate and make it the default implementation, otherwise it is going to be really annoying for all other embedders to embed Servo without a lot of code duplication.

wusyong commented on ports/servoshell/rendering_context.rs:
```diff
@@ -0,0 +1,297 @@
+}
+
+impl RenderingContext for SurfmanRenderingContext {
+    fn context(&self) -> Ref<'_, Context> {
+        self.0.context.borrow()
+    }
+
+    fn context_mut(&self) -> RefMut<'_, Context> {
+        self.0.context.borrow_mut()
+    }
+
+    fn device(&self) -> Ref<'_, Device> {
+        self.0.device.borrow()
+    }
+
+    fn device_mut(&self) -> RefMut<'_, Device> {
+        self.0.device.borrow_mut()
+    }
```
Could these just return & and &mut. I don't think other context will all use RefCell.

wusyong commented on components/shared/webrender/rendering_context.rs:
```diff
@@ -2,268 +2,17 @@
-    }
+use surfman::{Context, Device, Error, NativeWidget};
+
+pub trait RenderingContext {
```
We can add Clone trait bound here, so We don't need to pass Rc.

wusyong commented on components/servo/lib.rs:
```diff
@@ -236,7 +236,7 @@ where
         opts: Opts,
         preferences: Preferences,
-        rendering_context: RenderingContext,
+        rendering_context: Rc<dyn RenderingContext>,
```
Can this be a generic instead of trait object? I don't think we will hot reload the context after creation.

mrobinson commented on components/shared/webrender/rendering_context.rs:
```diff
@@ -189,31 +155,32 @@ impl RenderingContext {
         device.connection()
     }
 
+    #[allow(dead_code)]
```
What is the reasoning behind these new `dead_code` annotations?

dklassic replied:
They are probably not required now that `SurfmanRenderingContext` will remain in webrender crate, it's a byproduct from moving into servoshell. I'll do some reworking and will turn this PR into a draft state for now.

dklassic committed:
feat: turn RenderingContext into a trait

Signed-off-by: DK Liao <dklassic@gmail.com>

Add comment and handle drop error

Signed-off-by: DK Liao <dklassic@gmail.com>

Add doc comment

Signed-off-by: DK Liao <dklassic@gmail.com>

Allow make_current to propagate Error

Signed-off-by: DK Liao <dklassic@gmail.com>

Fix all make_current occurances

Signed-off-by: DK Liao <dklassic@gmail.com>

Move back to Rc<dyn RenderingContext>

Signed-off-by: DK Liao <dklassic@gmail.com>

fix android/ohos

Signed-off-by: DK Liao <dklassic@gmail.com>

Fix build

Signed-off-by: DK Liao <dklassic@gmail.com>

fix android/ohos again

Signed-off-by: DK Liao <dklassic@gmail.com>

dklassic commented:
I've reworked the trait methods and it should be cleaner now.

A minor discussion point is that `invalidate_native_surface` and `replace_native_surface` are both used in Android alone while being called using very specific `EmbedderEvent`. We should maybe move these work to be done within the ports themselves instead of having the compositor to handle them.

wusyong commented on components/canvas/webgl_mode/inprocess.rs:
```diff
@@ -125,6 +143,12 @@ impl WebGLExternalImages {
 
+impl Drop for WebGLExternalImages {
+    fn drop(&mut self) {
+        let _ = self.device.destroy_context(&mut self.context);
```
Could you log the error instead of ignoring here?

wusyong commented on components/shared/webrender/rendering_context.rs:
```diff
@@ -9,17 +9,39 @@ use std::ffi::c_void;
     SurfaceTexture, SurfaceType,
 };
 
+pub trait RenderingContext {
+    fn device(&self) -> NativeDevice;
+    fn context(&self) -> NativeContext;
+    fn resize(&self, size: Size2D<i32>);
+    fn present(&self);
+    fn bind_native_surface_to_context(&self, native_widget: NativeWidget);
+    fn connection(&self) -> Connection;
+    fn adapter(&self) -> Adapter;
+    fn make_current(&self);
+    fn framebuffer_object(&self) -> u32;
+    fn gl_api(&self) -> Rc<dyn gleam::gl::Gl>;
+    fn gl_version(&self) -> GLVersion;
+    fn invalidate_native_surface(&self);
+    fn replace_native_surface(
+        &self,
+        native_widget: *mut c_void,
+        coords: euclid::Size2D<i32, webrender_api::units::DevicePixel>,
+    );
+}
```
Could you add the documentation for this trait and its methods? You can take #34888 for some reference.

wusyong reviewed (approved):
This looks good to me. There are some trait methods I would remove or update, but that will affect how other components work. I suppose we could do follow up PR to handle them.

wusyong commented:
@mrobinson Could you review this too? Looks like it's blocked by your request change.

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -165,7 +165,7 @@ pub struct IOCompositor<Window: WindowMethods + ?Sized> {
 
     /// The surfman instance that webrender targets
-    rendering_context: RenderingContext,
+    rendering_context: R,
```
Instead of adding a new parameter to the structure, I think this should hold either `Box<dyn RenderingContext>` or `Rc<dyn RenderingContext>`.

dklassic replied:
Actually I changed it to generics because @wusyong asked me to, I was wondering about it too after seeing #35121

mrobinson replied:
@dklassic Apologies! I didn't notice this. Yes. We are starting to move away from making the `Compositor` generic on these types. Hopefully that is okay with you @wusyong.

Sorry for making you do and undo something.

wusyong replied:
@mrobinson  What's the reason to move away generic? Shouldn't generic have less overhead?

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -412,9 +412,7 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
-        if let Err(err) = self.rendering_context.make_gl_context_current() {
-            warn!("Failed to make GL context current: {:?}", err);
-        }
+        self.rendering_context.make_current();
```
We should probably allow the `RenderingContext` to propagate failure still so that the `Compositor` can do error handling. Just because it's only warning now doesn't mean it shouldn't handle errors. Quite likely it needs to stop itself and ask the application to restart it when this happens. None of that needs to happen here, but I think the `RenderingContext` should still expose the errors when they do happen.

mrobinson commented on components/compositing/lib.rs:
```diff
@@ -40,7 +40,7 @@ pub struct InitialCompositorState {
     pub webrender_document: DocumentId,
     pub webrender_api: RenderApi,
-    pub rendering_context: RenderingContext,
+    pub rendering_context: R,
```
Same comment about `dyn RenderingContext` here.

mrobinson commented on components/servo/lib.rs:
```diff
@@ -282,23 +284,14 @@ where
-                gl::GlesFns::load_with(|s| rendering_context.get_proc_address(s))
-            },
-        };
+        let webrender_gl = rendering_context.gl_api();
```
This is nice.

mrobinson reviewed (changes requested):
Looks pretty good, but I have just a few requests. Thanks for working on this!

dklassic commented:
https://github.com/dklassic/servo/actions/runs/12928709316/job/36056617147#step:16:70

Humm smoke test in macos showed some weird result, I should do some investigating.

dklassic commented:
> https://github.com/dklassic/servo/actions/runs/12928709316/job/36056617147#step:16:70
> 
> Humm smoke test in macos showed some weird result, I should do some investigation before we merge it.

A quick investigation showed that it’s probably due to GitHub runner not having access to Metal Device:

https://github.com/actions/runner-images/issues/7085

I’ll do more digging tmr and see if there’s a solution to this.

wusyong commented:
@dklassic Let's just pass `Rc<dyn RenderingContext>` to webgl image handler and create two trait methods create_texture/destroy_texture then. I'll think about how to update when I look into this image handler again.

dklassic committed:
Fix macOS smoke test

Signed-off-by: DK Liao <dklassic@gmail.com>

dklassic committed:
Add comment for create_texture and destroy_texture

Signed-off-by: DK Liao <dklassic@gmail.com>

dklassic committed:
Improve comments written

Signed-off-by: DK Liao <dklassic@gmail.com>

dklassic commented:
https://github.com/dklassic/servo/actions/runs/12934787333/job/36076635102

Okay it can pass the macOS DMG smoke test now. Not quite satisfied with the setup but I guess this the the best we can do for now.
